File: ABC/351-400/368/d.cpp
#include <bits/stdc++.h>
using namespace std;
#define rep(i,n) for(int i=0;i<n;++i)

int main(){
    int n,k;
    cin >> n >> k;
    vector<vector<int> > g(n);
    rep(i,n-1){
        int a,b;
        cin >> a >> b;
        a--; b--;
        g[a].push_back(b);
        g[b].push_back(a);
    }
    vector<int> vs(k);
    rep(i,k) cin >> vs[i],vs[i]--;
    vector<bool> connect(n);

    rep(i,k){
        if(i == 0) connect[vs[i]] = true;
        else{
            vector<bool> seen(n);
            seen[vs[i]] = true;
            vector<int> visited;
            auto dfs = [&](auto dfs, int v)->void{
                if(connect[v]){
                    for(int x : visited) connect[x] = true;
                    return;
                }
                seen[v] = true;
                visited.push_back(v);
                for(int u : g[v]){
                    if(seen[u]) continue;
                    dfs(dfs,u);
                    visited.pop_back();
                }
            };
            dfs(dfs,vs[i]);
        }
    }
    int ans = 0;
    rep(i,n) if(connect[i]) ans++;
    cout << ans << endl;
    return 0;
}

File: ABC/351-400/368/d.py
import sys
sys.setrecursionlimit(10**6)

n,k = map(int,input().split())
g = [[] for _ in range(n)]

for i in range(n-1):
    a,b = map(int,input().split())
    a-=1 
    b-=1
    g[a].append(b)
    g[b].append(a)

vs = list(map(int,input().split()))
vs = [v-1 for v in vs]
need = [False]*n
need[vs[0]] = True

for i in range(1,k):
    v = vs[i]
    if need[v]:
        continue
    seen = [False]*n
    seen[v] = True
    visited = []
    def dfs(v):
        if need[v]:
            for x in visited:
                need[x] = True
            return
        seen[v] = True
        visited.append(v)
        for u in g[v]:
            if seen[u]:
                continue
            dfs(u)
            visited.pop()
    dfs(v)

ans = 0

for i in range(n):
    if need[i]:
        ans += 1

print(ans)

def main():
    pass

if __name__ == "__main__":
    main()